src/entities/unsorted_state_adapter.ts

```diff
@@ -0,0 +1,175 @@
+): EntityStateAdapter<T>
+export function createUnsortedStateAdapter<T>(selectId: IdSelector<T>): any {
+  type R = EntityState<T>
+
+  function addOneMutably(entity: T, state: R): void
+  function addOneMutably(entity: any, state: any): void {
+    const key = selectIdValue(entity, selectId)
+
+    if (key in state.entities) {
+      return
+    }
+
+    state.ids.push(key)
+    state.entities[key] = entity
+  }
+
+  function addManyMutably(entities: T[], state: R): void
+  function addManyMutably(entities: any[], state: any): void {
+    for (const entity of entities) {
+      addOneMutably(entity, state)
+    }
+  }
+
+  function setAllMutably(entities: T[], state: R): void
+  function setAllMutably(entities: any[], state: any): void {
+    state.ids = []
+    state.entities = {}
+
+    addManyMutably(entities, state)
+  }
+
+  function removeOneMutably(key: T, state: R): void
+  function removeOneMutably(key: any, state: any): void {
+    return removeManyMutably([key], state)
+  }
+
+  function removeManyMutably(keys: any[], state: R): void {
+    let didMutate = false
+
+    keys.forEach(key => {
+      if (key in state.entities) {
+        delete state.entities[key]
+        didMutate = true
+      }
+    })
+
+    if (didMutate) {
+      // Work around TS not letting us call array methods if it's not a single known type
+      const ids = state.ids as string[]
+      state.ids = ids.filter((id: any) => id in state.entities)
+    }
+  }
+
+  function removeAll<S extends R>(state: S): S
+  function removeAll<S extends R>(state: any): S {
+    return Object.assign({}, state, {
+      ids: [],
+      entities: {}
+    })
+  }
+
+  function takeNewKey(
+    keys: { [id: string]: string },
+    update: Update<T>,
+    state: R
+  ): void
+  function takeNewKey(
+    keys: { [id: string]: any },
+    update: Update<T>,
+    state: any
+  ): boolean {
+    const original = state.entities[update.id]
+    const updated: T = Object.assign({}, original, update.changes)
+    const newKey = selectIdValue(updated, selectId)
+    const hasNewKey = newKey !== update.id
+
+    if (hasNewKey) {
+      keys[update.id] = newKey
+      delete state.entities[update.id]
+    }
+
+    state.entities[newKey] = updated
+
+    return hasNewKey
+  }
+
+  function updateOneMutably(update: Update<T>, state: R): void
+  function updateOneMutably(update: any, state: any): void {
+    return updateManyMutably([update], state)
+  }
+
+  function updateManyMutably(updates: Update<T>[], state: R): void
+  function updateManyMutably(updates: any[], state: any): void {
+    const newKeys: { [id: string]: string } = {}
+
+    updates = updates.filter(update => update.id in state.entities)
+
+    const didMutateEntities = updates.length > 0
+
+    if (didMutateEntities) {
+      const didMutateIds =
+        updates.filter(update => takeNewKey(newKeys, update, state)).length > 0
+
+      if (didMutateIds) {
+        state.ids = state.ids.map((id: any) => newKeys[id] || id)
```

Comment:
if two newKeys overlap (and we allow that), then one of the ids would be duplicated here, so `state.ids.length > Object.values(state.entities).length`

Reply:
Yeah, there's definitely a bug here.  Take this case:

```ts
    const withA = adapter.addOne(state, { id: 'a', title: 'First' })

    const withUpdates = adapter.updateMany(withA, [
      { id: 'a', changes: { id: 'b' } },
      { id: 'a', changes: { id: 'c' } }
    ])

    const { ids, entities } = withUpdates
```

The resulting state looks like:

```js
    {
      ids: [ 'c' ],
      entities: { b: { id: 'b', title: 'First' }, c: { id: 'c' } }
    }
```

The rename for B went through fully, but then the `'c'` rename didn't have an existing object to merge with.  The contents of just the update object were inserted, and we saved the `a -> c` ID mapping, so that's the final ID that ends up in `state.ids`

Reply:
Fixed it by changing the logic to merge together all updates targeting the same ID:

```ts
    const updatesPerEntity: { [id: string]: Update<T> } = {}

    updates.forEach(update => {
      // Only apply updates to entities that currently exist
      if (update.id in state.entities) {
        // If there are multiple updates to one entity, merge them together
        updatesPerEntity[update.id] = {
          ...updatesPerEntity[update.id],
          ...update
        }
      }
    })

    updates = Object.values(updatesPerEntity)
```

Reply:
and of course the spread operator _conveniently_ doesn't do anything with falsy values, allowing that line to work whether it's the tenth update for that item or the first.